### DISCUSSION TITLE
Tips for successfully compiling

### DISCUSSION THREAD

**Discussion by AverageHedgehog** [valinet/ExplorerPatcher#190] [General]
I've followed the guide for x64 after launching visual studio and building the release, it gives a ton of errors. Does the repository need to be in a specific location? I did everything in the documents folder.

**Comment by valinet** [comment]
Errors like…?

**Reply by AverageHedgehog** [reply]
I deleted the files and was going to try again today, first wanted to know which directory I should be working in or if it matters.

**Reply by valinet** [reply]
It doesn’t matter. Or it shouldn’t. Maybe there’s an issue there. Idk. Showing the errors is very useful. “Errors” doesn’t mean much from a debugging perspective.

**Reply by AverageHedgehog** [reply]
I agree, let me get some more info

**Comment by AverageHedgehog** [comment]
So building Release/amd64 gives this output error plus 500+ warnings.
![7wz0gjX](https://user-images.githubusercontent.com/93039050/138556801-da8eeaa2-82a4-42f3-aef2-655061d71045.png)

**Reply by AverageHedgehog** [reply]
This the general properties for visual studio
![yHYSmJE](https://user-images.githubusercontent.com/93039050/138557548-7cc236cc-d694-4c25-a418-88a76a001018.png)

**Reply by valinet** [reply]
Ignore the warnings... =))) Just tried now, it works for me. Which Windows SDK have you installed? If a newer one, maybe remove lines 21-57 from `utility.h` and try again?

**Reply by valinet** [reply]
Those just manually define some newer constants from`dwmapi.h` so that it compiles with older SDKs as well, I personally compile against 19041 because I've had this for a long time and it doesn't really require the newer one. I haven't looked how they implement those names, if they're in some `const enum`, maybe that's why it breaks. Idk, those should be, in `utility.h` only when compiling with older SDKs.

**Reply by valinet** [reply]
Maybe use a better guard there, like `_WIN32_WINNT_WIN11`. Does that exist in your SDK?

**Reply by AverageHedgehog** [reply]
I believe I have 10.0.22000 installed, I did not select it when installing visual studio and used the installer.

**Reply by valinet** [reply]
If you click the drop down where you select "Windows SDK Version" in project properties it lists you exactly the versions installed. So yeah, did removing those lines allow you to compile?

**Reply by AverageHedgehog** [reply]
No, just gave an additional error. I think you can manually switch the SDK version so I will try again and select 19041
![hkXNfXY](https://user-images.githubusercontent.com/93039050/138558477-0ec91e24-0ff2-4299-95e1-52fb873c7992.png)

**Reply by valinet** [reply]
Well, leave in the `DWMWA_MICA_EFFECT`, that indeed was custom defined, they still haven't documented those. Double click on the first error and show me what it underlines you in the file. You don't need to waste time installing a different SDK, it works with this as well, let's fix those 2 errors...

**Reply by valinet** [reply]
They maybe just fucked up a bit the definitions from being used in C.

**Reply by AverageHedgehog** [reply]
Changing the SDK and reverting utility.h gives this 
![image](https://user-images.githubusercontent.com/93039050/138558698-b88dccbd-f143-4778-8dd7-89397f1aa93c.png)

**Reply by AverageHedgehog** [reply]
> Well, leave in the `DWMWA_MICA_EFFECT`, that indeed was custom defined, they still haven't documented those. Double click on the first error and show me what it underlines you in the file. You don't need to waste time installing a different SDK, it works with this as well, let's fix those 2 errors...

 DWMWA_USE_HOSTBACKDROPBRUSH,                // [set] BOOL, Allows the use of host backdrop brushes for the window.

and 

/*
     * Let the system decide whether or not to round window corners
     */
    DWMWCP_DEFAULT                                 = 0,

**Reply by valinet** [reply]
Why can't you listen to what I say...? Maybe `funchook` now needs recompiled as well for the older SDK. But just DON'T DO THAT. LET'S FIX THE ERRORS FROM THE FIRST TRY, no need to install a million SDKs, whatever you had initially is fine, let's fix it on that.

**Reply by valinet** [reply]
> 
> 
> > Well, leave in the `DWMWA_MICA_EFFECT`, that indeed was custom defined, they still haven't documented those. Double click on the first error and show me what it underlines you in the file. You don't need to waste time installing a different SDK, it works with this as well, let's fix those 2 errors...
> 
> DWMWA_USE_HOSTBACKDROPBRUSH, // [set] BOOL, Allows the use of host backdrop brushes for the window.
> 
> and
> 
> /* * Let the system decide whether or not to round window corners */ DWMWCP_DEFAULT = 0,

Show me how this looks in the `dwmapi.h` file from 22000 whatever:

```
// Window attributes
enum DWMWINDOWATTRIBUTE
{
    DWMWA_NCRENDERING_ENABLED = 1,      // [get] Is non-client rendering enabled/disabled
    DWMWA_NCRENDERING_POLICY,           // [set] DWMNCRENDERINGPOLICY - Non-client rendering policy
    DWMWA_TRANSITIONS_FORCEDISABLED,    // [set] Potentially enable/forcibly disable transitions
    DWMWA_ALLOW_NCPAINT,                // [set] Allow contents rendered in the non-client area to be visible on the DWM-drawn frame.
    DWMWA_CAPTION_BUTTON_BOUNDS,        // [get] Bounds of the caption button area in window-relative space.
    DWMWA_NONCLIENT_RTL_LAYOUT,         // [set] Is non-client content RTL mirrored
    DWMWA_FORCE_ICONIC_REPRESENTATION,  // [set] Force this window to display iconic thumbnails.
    DWMWA_FLIP3D_POLICY,                // [set] Designates how Flip3D will treat the window.
    DWMWA_EXTENDED_FRAME_BOUNDS,        // [get] Gets the extended frame bounds rectangle in screen space
    DWMWA_HAS_ICONIC_BITMAP,            // [set] Indicates an available bitmap when there is no better thumbnail representation.
    DWMWA_DISALLOW_PEEK,                // [set] Don't invoke Peek on the window.
    DWMWA_EXCLUDED_FROM_PEEK,           // [set] LivePreview exclusion information
    DWMWA_CLOAK,                        // [set] Cloak or uncloak the window
    DWMWA_CLOAKED,                      // [get] Gets the cloaked state of the window
    DWMWA_FREEZE_REPRESENTATION,        // [set] BOOL, Force this window to freeze the thumbnail without live update
    DWMWA_PASSIVE_UPDATE_MODE,          // [set] BOOL, Updates the window only when desktop composition runs for other reasons
    DWMWA_LAST
};
```

**Reply by AverageHedgehog** [reply]
Changed back to the latest SDK gives this:
![image](https://user-images.githubusercontent.com/93039050/138558842-dadb792a-3e6e-4733-bf4b-7a0efd0abed7.png)

**Reply by AverageHedgehog** [reply]
```
// Window attributes
enum DWMWINDOWATTRIBUTE
{
    DWMWA_NCRENDERING_ENABLED = 1,              // [get] Is non-client rendering enabled/disabled
    DWMWA_NCRENDERING_POLICY,                   // [set] DWMNCRENDERINGPOLICY - Non-client rendering policy
    DWMWA_TRANSITIONS_FORCEDISABLED,            // [set] Potentially enable/forcibly disable transitions
    DWMWA_ALLOW_NCPAINT,                        // [set] Allow contents rendered in the non-client area to be visible on the DWM-drawn frame.
    DWMWA_CAPTION_BUTTON_BOUNDS,                // [get] Bounds of the caption button area in window-relative space.
    DWMWA_NONCLIENT_RTL_LAYOUT,                 // [set] Is non-client content RTL mirrored
    DWMWA_FORCE_ICONIC_REPRESENTATION,          // [set] Force this window to display iconic thumbnails.
    DWMWA_FLIP3D_POLICY,                        // [set] Designates how Flip3D will treat the window.
    DWMWA_EXTENDED_FRAME_BOUNDS,                // [get] Gets the extended frame bounds rectangle in screen space
    DWMWA_HAS_ICONIC_BITMAP,                    // [set] Indicates an available bitmap when there is no better thumbnail representation.
    DWMWA_DISALLOW_PEEK,                        // [set] Don't invoke Peek on the window.
    DWMWA_EXCLUDED_FROM_PEEK,                   // [set] LivePreview exclusion information
    DWMWA_CLOAK,                                // [set] Cloak or uncloak the window
    DWMWA_CLOAKED,                              // [get] Gets the cloaked state of the window
    DWMWA_FREEZE_REPRESENTATION,                // [set] BOOL, Force this window to freeze the thumbnail without live update
    DWMWA_PASSIVE_UPDATE_MODE,                  // [set] BOOL, Updates the window only when desktop composition runs for other reasons
    DWMWA_USE_HOSTBACKDROPBRUSH,                // [set] BOOL, Allows the use of host backdrop brushes for the window.
    DWMWA_USE_IMMERSIVE_DARK_MODE = 20,         // [set] BOOL, Allows a window to either use the accent color, or dark, according to the user Color Mode preferences.
    DWMWA_WINDOW_CORNER_PREFERENCE = 33,        // [set] WINDOW_CORNER_PREFERENCE, Controls the policy that rounds top-level window corners
    DWMWA_BORDER_COLOR,                         // [set] COLORREF, The color of the thin border around a top-level window
    DWMWA_CAPTION_COLOR,                        // [set] COLORREF, The color of the caption
    DWMWA_TEXT_COLOR,                           // [set] COLORREF, The color of the caption text
    DWMWA_VISIBLE_FRAME_BORDER_THICKNESS,       // [get] UINT, width of the visible border around a thick frame window
    DWMWA_LAST
};
```

**Reply by valinet** [reply]
Just upload the `dwmapi.h` file from 22000 here, please.

**Reply by AverageHedgehog** [reply]
[dwmapi.txt](https://github.com/valinet/ExplorerPatcher/files/7403094/dwmapi.txt)

**Reply by valinet** [reply]
Find the file `sws_def.h` and remove these from in there as well:

```
#define DWMWA_USE_HOSTBACKDROPBRUSH 17            // [set] BOOL, Allows the use of host backdrop brushes for the window.
#define DWMWA_USE_IMMERSIVE_DARK_MODE 20          // [set] BOOL, Allows a window to either use the accent color, or dark, according to the user Color Mode preferences.
#define DWMWA_WINDOW_CORNER_PREFERENCE 33         // [set] WINDOW_CORNER_PREFERENCE, Controls the policy that rounds top-level window corners
#define DWMWA_BORDER_COLOR 34                     // [set] COLORREF, The color of the thin border around a top-level window
#define DWMWA_CAPTION_COLOR 35                    // [set] COLORREF, The color of the caption
#define DWMWA_TEXT_COLOR 36                       // [set] COLORREF, The color of the caption text
#define DWMWA_VISIBLE_FRAME_BORDER_THICKNESS 37   // [get] UINT, width of the visible border around a thick frame window
```

```
#define DWMWCP_DEFAULT 0
#define DWMWCP_DONOTROUND 1
#define DWMWCP_ROUND 2
#define DWMWCP_ROUNDSMALL 3
```

That error just tells you that those have already been defined in the `dwmapi.h` file, as is the case on newer SDKs. On older SDKs, I had to define those manually, indeed.

**Reply by AverageHedgehog** [reply]
I removed those lines from sws_def.h and I'm getting this now
![J3blcGR](https://user-images.githubusercontent.com/93039050/138560214-c2c2279a-c3f7-4eee-a0d7-e07fafd42f05.png)
:

**Reply by valinet** [reply]
Yeah, `funchook` has to be linked against the static runtime library. From `README`:

   Now, in the `libs\funchook\build` folder, open the file `funchook-static.vcxproj` with any text editor, search and replace all occurences of `<RuntimeLibrary>MultiThreadedDLL</RuntimeLibrary>` with `<RuntimeLibrary>MultiThreaded</RuntimeLibrary>`.

That's the cause. Idk, do that, or change EP to link against MD instead of MT in Project properties -> C/C++ Compiler.

**Reply by valinet** [reply]
After you edit that file, recompile funchook.

**Reply by AverageHedgehog** [reply]
> After you edit that file, recompile funchook.

I fixed funchook and recompiled it, now EP is able to build in VS. But when running the command prompt I get this. So close lol, thanks for helping with this and I hope it's helping you.
![4Pvm9LK](https://user-images.githubusercontent.com/93039050/138562238-a6c9d646-969a-463b-ba3a-973c1d47c3ec.png)

**Reply by valinet** [reply]
Compiling from terminal using MSBuild is the same thing VS does anyway. So idk why you need to do that as well but whatever. The error is probably because there is no "x64" platform anymore in my solution file, I renamed it to "amd64". So just use:

```
msbuild ExplorerPatcher.sln /property:Configuration=Release /property:Platform=amd64
```

**Reply by AverageHedgehog** [reply]
That seemed to work. I don't see an exe where specified though, can you please explain the final install steps?

![image](https://user-images.githubusercontent.com/93039050/138563011-10923b26-dfe7-4b58-8761-cc1dd2d17ff6.png)

**Reply by valinet** [reply]
What EXE file? This compiles as a library. It generates an `ExplorerPatcher.amd64.dll` file somewhere...

**Reply by valinet** [reply]
Again, compiling from terminal using MSBuild or from Visual Studio is the same thing. In your image, it anyway did not bother doing much anyway, since it found everything was up to date from your previous compilation in VS. Just use VS if you have that... the command line instructions are posted there just in case.

**Reply by AverageHedgehog** [reply]
Ah okay, I miss read the readme then, "The resulting exe and dll will be in "Release" folder (if you chose x86), or "x64\Release" (if you chose x64) in the folder containing the solution file." Seems to be installed, thank you.

**Reply by valinet** [reply]
The files are most probably created in a "build" directory in the root folder. Again, changed that recently, the README may be outdated.

**Reply by valinet** [reply]
> 
> 
> Ah okay, I miss read the readme then, "The resulting exe and dll will be in "Release" folder (if you chose x86), or "x64\Release" (if you chose x64) in the folder containing the solution file." Seems to be installed, thank you.

Yeah, it's outdated, that part is copy pasted from other projects I have... apologies as well. But like, you know, if I ship only DLLs in releases, what EXE would it create if compiled manually...?

**Reply by AverageHedgehog** [reply]
:)
![image](https://user-images.githubusercontent.com/93039050/138563210-a4018688-cc9b-4372-972d-ff85d9e8821e.png)

**Reply by valinet** [reply]
So, again, sorry for the slightly outdated docs or if I were a bit too rushing/sour in my explanations.
